File: lc567.cs
public class Solution {
    public bool CheckInclusion(string s1, string s2) {
        Dictionary<char, int> map = new Dictionary<char, int>();
        foreach(char c in s1)
        {
            if(!map.ContainsKey(c))
            {
                map.Add(c,0);
            }
            map[c]++;
        }
        Dictionary<char, int> smap = new Dictionary<char, int>();
        for(int i=0;i<s2.Length;i++)
        {
            if(!smap.ContainsKey(s2[i]))
            {
                smap.Add(s2[i],0);
            }            
            smap[s2[i]]++;
            if(i>=s1.Length-1)
            {
                if(CheckIfMapsAreEqual(map, smap))
                {
                    return true;
                }
                smap[s2[i-s1.Length+1]]--;
                if(smap[s2[i-s1.Length+1]]==0)
                {
                    smap.Remove(s2[i-s1.Length+1]);
                }
            }
        }
        return false;
    }
    private bool CheckIfMapsAreEqual(Dictionary<char, int> a, Dictionary<char, int> b)
    {
        if(a.Count != b.Count)
            return false;
        foreach(var kvp in a)
        {
            if(b.ContainsKey(kvp.Key) && a[kvp.Key] ==b[kvp.Key])
            {
                //we are good
            }
            else
            {
                return false;
            }
        }
        return true;
    }    
}

File: java/lc567.java
class Solution {
    public boolean checkInclusion(String p, String s) {        
        HashMap<Character, Integer> map = new HashMap<Character, Integer>();
        for(char c : p.toCharArray())
        {
            if(!map.containsKey(c))
            {
                map.put(c,0);
            }
            map.put(c,map.get(c)+1);
        }
        HashMap<Character, Integer> smap = new HashMap<Character, Integer>();
        for(int i=0;i<s.length();i++)
        {
            if(!smap.containsKey(s.charAt(i)))
            {
                smap.put(s.charAt(i),1);
            }            
            else
            {
                smap.put(s.charAt(i),map.get(s.charAt(i))+1);
            }
            if(i>=p.length()-1)
            {
                if(CheckIfMapsAreEqual(map, smap))
                {
                   return true;
                }
                char ch = s.charAt(i-p.length()+1);
                int val = smap.get(ch);
                smap.put(ch,val-1);
                if(smap.get(ch)==0)
                {
                    smap.remove(ch);
                }
            }
        }
        return false;
    }
    private boolean CheckIfMapsAreEqual(HashMap<Character, Integer> a, HashMap<Character, Integer> b)
    {
        if(a.size() != b.size())
            return false;
        for(char c: a.keySet())
        {
            if(b.containsKey(c) && a.get(c) ==b.get(c))
            {
                //we are good
            }
            else
            {
                return false;
            }
        }
        return true;
    }    
}
